cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/RackAwareDistributionGoal.java

```diff
@@ -0,0 +1,353 @@
+    GoalUtils.ensureNoOfflineReplicas(clusterModel, name());
+    // One pass is sufficient to satisfy or alert impossibility of this goal.
+    // Sanity check to confirm that replicas of each partition are evenly distributed across the racks
+    ensureRackAwareDistribution(clusterModel, optimizationOptions);
+    // Sanity check: No replica should be moved to a broker, which used to host any replica of the same partition on its broken disk.
+    GoalUtils.ensureReplicasMoveOffBrokersWithBadDisks(clusterModel, name());
+    finish();
+  }
+
+  @Override
+  public void finish() {
+    super.finish();
+    // Clean up the memory
+    _balanceLimit.clear();
+  }
+
+  /**
+   * Violations of rack-aware distribution can be resolved with replica movements.
+   *
+   * @param broker Broker to be balanced.
+   * @param clusterModel The state of the cluster.
+   * @param optimizedGoals Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker,
+                                    ClusterModel clusterModel,
+                                    Set<Goal> optimizedGoals,
+                                    OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    rebalanceForBroker(broker, clusterModel, optimizedGoals, optimizationOptions, false);
+  }
+
+  /**
+   * Get a list of eligible brokers for moving the given replica in the given cluster to satisfy the specific
+   * requirements of the custom rack aware goal.
+   * A broker is rack aware distribution eligible for a given replica if it is in a rack
+   * <ul>
+   *   <li>that needs more replicas -- i.e. has fewer than base limit replicas</li>
+   *   <li>with at most as many as base limit replicas, and the number of racks with at least one more replica over the
+   *   base limit is fewer than the allowed number of such racks</li>
+   * </ul>
+   *
+   * @param replica Replica for which a set of rack aware eligible brokers are requested.
+   * @param clusterModel The state of the cluster.
+   * @return A list of rack aware eligible brokers for the given replica in the given cluster.
+   */
+  @Override
+  protected SortedSet<Broker> rackAwareEligibleBrokers(Replica replica, ClusterModel clusterModel) {
+    Set<Broker> partitionBrokers = clusterModel.partition(replica.topicPartition()).partitionBrokers();
+    Map<String, Integer> numReplicasByRack = numPartitionReplicasByRackId(partitionBrokers);
+
+    // Decrement the number of replicas in this rack.
+    numReplicasByRack.merge(replica.broker().rack().id(), -1, Integer::sum);
+
+    int baseLimit = _balanceLimit.baseLimitByRF(partitionBrokers.size());
+
+    // An replica is allowed to be moved to a rack at the base limit only if the number of racks with at least one more
+    // replica over the base limit is fewer than the allowed number of such racks
```

Comment:
Can an example be added in the comments that illustrates this case?

Reply:
Sure -- added an example.